internal/config/variables/variables.go

```diff
@@ -0,0 +1,531 @@
+	// of env, file, cli.
+
+	// process env values ("env" source)
+	{
+		env := os.Environ()
+		for _, raw := range env {
+			if !strings.HasPrefix(raw, varEnvPrefix) {
+				continue
+			}
+			raw = raw[len(varEnvPrefix):] // trim the prefix
+
+			eq := strings.Index(raw, "=")
+			if eq == -1 {
+				// Seems invalid, so we'll ignore it.
+				continue
+			}
+
+			name := raw[:eq]
+			rawVal := raw[eq+1:]
+
+			ret = append(ret, &pb.Variable{
+				Name:   name,
+				Value:  &pb.Variable_Str{Str: rawVal},
+				Source: &pb.Variable_Env{},
+			})
+		}
+	}
+
+	// process -var-file args ("file" source)
+	for _, file := range files {
+		if file != "" {
+			pbv, diags := parseFileValues(file, sourceFile)
+			if diags.HasErrors() {
+				return nil, diags
+			}
+			ret = append(ret, pbv...)
+		}
+	}
+
+	// process -var args ("cli" source)
+	for name, val := range vars {
+		ret = append(ret, &pb.Variable{
+			Name:   name,
+			Value:  &pb.Variable_Str{Str: val},
+			Source: &pb.Variable_Cli{},
+		})
+	}
+	return ret, diags
+}
+
+// EvalInputValues evaluates the provided variable values and validates their
+// types per the type declared in the waypoint.hcl for that variable name.
+// The order in which values are evaluated corresponds to their precedence, with
+// higher precedence values overwriting lower precedence values.
+func EvalInputValues(
+	pbvars []*pb.Variable,
+	vs map[string]*Variable,
+) (InputValues, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	iv := InputValues{}
+
+	for v, def := range vs {
+		iv[v] = def.Default
+	}
+
+	for _, pbv := range pbvars {
+		variable, found := vs[pbv.Name]
+		if !found {
+			// TODO krantzinator: this should be a DiagWarning rather than a DiagErr
+			// What to do with a warning diag type?
+			diags = append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  "Undefined variable",
+				Detail: fmt.Sprintf("A %q variable value was set, "+
+					"but was not found in known variables. To declare variable "+
+					"%q, place a variable definition block in your waypoint.hcl file.",
+					pbv.Name, pbv.Name),
+			})
+			continue
+		}
+
+		var source string
+		switch pbv.Source.(type) {
+		case *pb.Variable_Cli:
+			source = sourceCLI
+		case *pb.Variable_File_:
+			source = sourceFile
+		case *pb.Variable_Env:
+			source = sourceEnv
+		case *pb.Variable_Vcs:
+			source = sourceVCS
+		case *pb.Variable_Server:
+			source = sourceServer
+		}
+
+		// We have to specify the three different simple types we support -- string,
+		// bool, number -- when doing the below evaluation of hcl expressions
+		// because of our translation to-and-from protobuf format.
+		// While cty allows us to parse all simple types as LiteralValueExpr, we
+		// have to first translate the pb values back into cty values, thus
+		// necessitating a separate case statement for each simple type
+		var expr hclsyntax.Expression
+		switch pbv.Value.(type) {
```

Comment:
This is a minor nit, you could use `switch sv := pbv.Value.(type)` here and then below, you can use `sv` as the type that was detected, so like on line 287, you could have `sv` instead of `pbv.Value(*pb.Variable_Hcl)`.

Reply:
I'd say this isn't a nit and is pretty important, this gets you much better type safety imo.

Reply:
Done!